Add pickling support for WorkerInfo

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73373 [Dynamic RPC] Allow newly joined ranks to communicate with existing ranks
* #73372 [Dynamic RPC] Allow for optional world_size argument in init_rpc
* **#73371 Add pickling support for WorkerInfo**

This PRs allows for the pybinded class `WorkerInfo` to be pickled. The class is pickled into a tuple of worker_name and rank in format `(NAME, ID)`. This allows WorkerInfo to be passed as an argument for RPC calls.

Differential Revision: [D34458153](https://our.internmc.facebook.com/intern/diff/D34458153)

H-Huang committed:
Add pickling support for WorkerInfo

[ghstack-poisoned]

H-Huang committed:
Update on "Add pickling support for WorkerInfo"


This PRs allows for the pybinded class WorkerInfo to be pickled. The class is pickled into a tuple of worker_name and rank in format `(NAME, ID)`. This allows WorkerInfo to be passed as an argument for RPC calls.

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34458153).

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -110,11 +110,26 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+              },
+              [](py::tuple t) { // __setstate__
+                if (t.size() != 2) {
+                  throw std::runtime_error("Invalid state");
```
use `TORCH_CHECK`?

H-Huang replied:
Thanks! Changed it

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -110,11 +110,26 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                WorkerInfo workerInfo(
+                    t[0].cast<std::string>(), t[1].cast<worker_id_t>());
+                return workerInfo;
+              }));
```
Let's add some tests for these?

H-Huang replied:
Thanks :) Added a test in rpc_test to do a simple rpc_sync call using WorkerInfo to test this.

mrshenli reviewed (approved):
LGTM

H-Huang committed:
Update on "Add pickling support for WorkerInfo"


This PRs allows for the pybinded class `WorkerInfo` to be pickled. The class is pickled into a tuple of worker_name and rank in format `(NAME, ID)`. This allows WorkerInfo to be passed as an argument for RPC calls.

Differential Revision: [D34458153](https://our.internmc.facebook.com/intern/diff/D34458153)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34458153).
